sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2725,4 +2725,14 @@ class Dataset[T] private[sql](
   @inline private def withTypedPlan[U : Encoder](logicalPlan: => LogicalPlan): Dataset[U] = {
     Dataset(sparkSession, logicalPlan)
   }
+
+  /** A convenient function to wrap a set based logical plan and produce a Dataset. */
+  @inline private def withSetOperator[U : Encoder](logicalPlan: => LogicalPlan): Dataset[U] = {
+    if (classTag.runtimeClass == classOf[Row]) {
```

Comment:
I _guess_ it'd be fine within Spark as it always set `clsTag` to `Row` via [RowEncoder.scala#L59]( https://github.com/apache/spark/blob/39e2bad6a866d27c3ca594d15e574a1da3ee84cc/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala#L59) in the path creating `DataFrame`; however, it seems potentially problematic if a custom encoder is used with a subclass of `Row` as I guess you meant.

Let me try to handle this. Thank you.